File: frameworks/motia/motia-js/packages/motia/src/triggers.ts
import type {
  ApiMiddleware,
  ApiRouteMethod,
  ApiTrigger,
  CronTrigger,
  QueryParam,
  QueueConfig,
  QueueTrigger,
  StateTrigger,
  StepSchemaInput,
  StreamTrigger,
  TriggerCondition,
} from './types'

type ApiOptions<TSchema extends StepSchemaInput | undefined = undefined> = {
  bodySchema?: TSchema
  responseSchema?: Record<number, StepSchemaInput>
  queryParams?: readonly QueryParam[]
  // biome-ignore lint/suspicious/noExplicitAny: we need to define this type to avoid type errors
  middleware?: readonly ApiMiddleware<any, any, any>[]
}

type QueueOptions<TSchema extends StepSchemaInput | undefined = undefined> = {
  input?: TSchema
  config?: Partial<QueueConfig>
}

// biome-ignore lint/suspicious/noExplicitAny: we need any to accept all schema types
export function http<TOptions extends ApiOptions<any> | undefined = undefined>(
  method: ApiRouteMethod,
  path: string,
  options?: TOptions,
  condition?: TriggerCondition,
): ApiTrigger<TOptions extends ApiOptions<infer S> ? S : undefined> {
  // biome-ignore lint/suspicious/noExplicitAny: runtime return is correct, cast needed for flexible type
  return { type: 'http', method, path, ...options, condition } as any
}

/** @deprecated Use http() instead. Will be removed in a future version. */
export function api<TOptions extends ApiOptions<any> | undefined = undefined>(
  method: ApiRouteMethod,
  path: string,
  options?: TOptions,
  condition?: TriggerCondition,
): ApiTrigger<TOptions extends ApiOptions<infer S> ? S : undefined> {
  return http(method, path, options, condition)
}

// biome-ignore lint/suspicious/noExplicitAny: we need any to accept all schema types
export function queue<TOptions extends QueueOptions<any> | undefined = undefined>(
  topic: string,
  options?: TOptions,
  condition?: TriggerCondition,
): QueueTrigger<TOptions extends QueueOptions<infer S> ? S : undefined> {
  // biome-ignore lint/suspicious/noExplicitAny: runtime return is correct, cast needed for flexible type
  return { type: 'queue', topic, ...options, condition } as any
}

export function cron(expression: string, condition?: TriggerCondition): CronTrigger {
  return { type: 'cron', expression, condition }
}

export function state(condition?: TriggerCondition): StateTrigger {
  return { type: 'state', condition }
}

type StreamOptions = {
  groupId?: string
  itemId?: string
  condition?: TriggerCondition
}

export function stream(streamName: string, optionsOrCondition?: StreamOptions | TriggerCondition): StreamTrigger {
  if (typeof optionsOrCondition === 'function') {
    return { type: 'stream', streamName, condition: optionsOrCondition }
  }
  const { groupId, itemId, condition } = optionsOrCondition ?? {}
  return { type: 'stream', streamName, groupId, itemId, condition }
}

File: frameworks/motia/motia-py/packages/motia/src/motia/triggers.py
"""Trigger helper constructors."""

from __future__ import annotations

import warnings
from typing import Any

from .types import ApiRouteMethod, ApiTrigger, CronTrigger, QueueTrigger, StateTrigger, StreamTrigger, TriggerCondition


def http(
    method: ApiRouteMethod,
    path: str,
    *,
    body_schema: Any | None = None,
    response_schema: dict[int, Any] | None = None,
    query_params: list[Any] | None = None,
    middleware: list[Any] | None = None,
    condition: Any | None = None,
) -> ApiTrigger:
    """Create an HTTP trigger configuration."""
    return ApiTrigger(
        path=path,
        method=method,
        condition=condition,
        body_schema=body_schema,
        response_schema=response_schema,
        query_params=query_params,
        middleware=middleware,
    )


def api(
    method: ApiRouteMethod,
    path: str,
    *,
    body_schema: Any | None = None,
    response_schema: dict[int, Any] | None = None,
    query_params: list[Any] | None = None,
    middleware: list[Any] | None = None,
    condition: Any | None = None,
) -> ApiTrigger:
    warnings.warn("api() is deprecated, use http() instead", DeprecationWarning, stacklevel=2)
    return http(
        method=method,
        path=path,
        body_schema=body_schema,
        response_schema=response_schema,
        query_params=query_params,
        middleware=middleware,
        condition=condition,
    )


def queue(
    topic: str,
    *,
    input: Any | None = None,
    config: Any | None = None,
    condition: Any | None = None,
) -> QueueTrigger:
    """Create a queue trigger configuration."""
    return QueueTrigger(
        topic=topic,
        condition=condition,
        input=input,
        config=config,
    )


def cron(expression: str, *, condition: Any | None = None) -> CronTrigger:
    """Create a cron trigger configuration."""
    return CronTrigger(expression=expression, condition=condition)


def state(*, condition: TriggerCondition | None = None) -> StateTrigger:
    """Create a state trigger."""
    return StateTrigger(type="state", condition=condition)


def stream(
    stream_name: str,
    *,
    group_id: str | None = None,
    item_id: str | None = None,
    condition: TriggerCondition | None = None,
) -> StreamTrigger:
    """Create a stream trigger."""
    return StreamTrigger(
        type="stream",
        stream_name=stream_name,
        group_id=group_id,
        item_id=item_id,
        condition=condition,
    )
